### DISCUSSION TITLE
CHILD SA states on receiving UP/DOWN events

### DISCUSSION THREAD

**Discussion by Pankaj-Deshpande-Com** [strongswan/strongswan#1228] [Q&A]
[URGENT QUERY] 

Hello,

In our VICI API based application, we are doing only reference counting for child SAs rather than remembering the full identifiers etc - this is to save memory and make our design scalable.

- We found possible CHILD SA specific states getting reported to VICI application are : "CREATED", "ROUTED", "INSTALLING", "INSTALLED", "UPDATING", "REKEYING", "REKEYED", "RETRYING", "DELETING", "DELETED", "DESTROYING"

- In our VICI based Application, we are doing ChildSARefCnt++ whenever we receive CHILD SA UP event and We are doing ChildSARefCnt-- whenever we receive CHILD SA DOWN event. But this may not be correct as Strongswan might sent two UP events in creation path in certain situations, in certain paths e.g. UP(CREATED)-->UP(INSTALLED) but in destroy path, it may give us single DOWN (may be DELETED) event. And there may be many more such combinations in which ref counting can go wrong/awry.

- We are keen to know what states we should use for reference counting in UP and DOWN path
    e.g.Take reference when we receive CHILD SA with state=CREATED and Down the reference when we receive CHILD SA DOWN with state=DELETED ?

So, What states reported in UP event and in what all states reported with DOWN event should we hook our reference count logic should we attach to ?
    and exactly how ? You had helped us solve the issue related to IKE SA UP/DOWN events and now we need your urgent help on CHILD SA events...

Regards
Pankaj Deshpande

**Comment by tobiasbrunner** [comment]
> But this may not be correct as Strongswan might sent two UP events in creation path in certain situations, in certain paths e.g. UP(CREATED)-->UP(INSTALLED) but in destroy path,

I don't think that's true. There should be only one UP event (_child-updown_ with _up=yes_) for every CHILD_SA (and a single DOWN event, _child-updown_ without _up_). Are you maybe confusing this with state changes (but those are currently not actually exposed via vici) or are you using some other event?

> Take reference when we receive CHILD SA with state=CREATED and Down the reference when we receive CHILD SA DOWN with state=DELETED ?

I don't think there is ever an event for CREATED (the SA might be in that state for other events, but never in the _child-updown_ event). There are state change events for INSTALLED (or ROUTED) and eventually for DESTROYING, but the latter is triggered for every CHILD_SA that is ever created and destroyed (even if it is never installed). So refcounting based on that wouldn't be a good idea (tracking the SA's unique IDs in a hashtable that way would probably work, though). But again those state changes are currently not exposed via vici.

> So, What states reported in UP event and in what all states reported with DOWN event should we hook our reference count logic should we attach to ?

The _child-updown_ event.

**Reply by Thermi** [reply]
> I don't think that's true. There should be only one UP event (child-updown with up=yes) for every CHILD_SA (and a single DOWN event, child-updown without up). 

that's true for how the legacy updown script does things if there are several components in the traffic selector on either or both sides.
Probably not true for the internal bus.

EDIT: clarification/precision in regards to behaviour

**Reply by tobiasbrunner** [reply]
> Probably not true for the internal bus.

And neither for VICI (and its _child-updown_ event), which the OP is referring to.
